spec/formatting.md

```diff
@@ -677,8 +677,14 @@ These are divided into the following categories:
     > ```
 
   - A **_Missing Selector Annotation error_** is an error that occurs when the _message_
-    contains a _selector_ that does not have an _annotation_,
-    or contains a _variable_ that does not directly or indirectly reference a _declaration_ with an _annotation_.
+    contains a _selector_ without an _annotation_.
+
+    An implementation MUST NOT emit this error for a _selector_ _expression_
+    that would have an _annotation_
+    if each local _variable_ in it were replaced
+    with the _expression_ of its binding _declaration_,
+    repeating this process until the _expression_
+    contains no local _variable_s.
```

Comment:
This approach is a bit problematic because option values cannot be expressions, and this is clearly including their replacement as well.

@stasm This is an example of a situation where aligning our data model closely with our syntax is making it more difficult to speak about it. If we went the other way than proposed in #436 and e.g. defined a _variable expression_ in the data model, we could use that term here.

Reply:
@eemeli I see your point.

The validity property is easy to express in terms of types:

```
The type of a local _variable_ is the type of the _expression_ in its _declaration_.
The type of an external variable is `Unannotated`.

The type of an _expression_ is:
* `Annotated` if the _expression_ has an _annotation_.
* `Annotated` if the _operand_ of the expression is a _variable_ whose type is `Annotated`.
* `Unannotated` otherwise.

A `selector` must have type `Annotated`.
```

I'm not sure how this language would be received, but we are talking about a static dataflow property, whether or not we call it that :)

I also agree that it would be easier to write an unambiguous specification if the object language was the language described by the abstract syntax, rather than the concrete syntax.